ekncontent/ekncontent/eknc-engine.c

```diff
@@ -0,0 +1,499 @@
+                   gpointer user_data)
+{
+  g_return_if_fail (EKNC_IS_ENGINE (self));
+  g_return_if_fail (EKNC_IS_QUERY_OBJECT (query));
+  g_return_if_fail (G_IS_CANCELLABLE (cancellable) || cancellable == NULL);
+
+  g_autofree gchar *query_app_id = NULL;
+  g_object_get (G_OBJECT (query),
+                "app_id", &query_app_id,
+                NULL);
+  gchar *app_id = self->default_app_id;
+  if (query_app_id && *query_app_id)
+    app_id = query_app_id;
+
+  g_return_if_fail (app_id && *app_id);
+
+  g_autoptr(GTask) task = g_task_new (self, cancellable, callback, user_data);
+  GError *error;
+  EkncDomain *domain;
+  if (!(domain = eknc_engine_get_domain_for_app (self, app_id, &error)))
+    {
+      g_task_return_error (task, error);
+      return;
+    }
+
+  g_autofree gchar *query_string = NULL;
+  g_object_get (query, "query", &query_string, NULL);
+  if (query_string && *query_string)
+    {
+      eknc_domain_get_fixed_query (domain, query, cancellable, on_domain_fix_query_finished, g_steal_pointer (&task));
+      return;
+    }
+
+  eknc_domain_query (domain, query, cancellable, on_domain_query_finished, g_steal_pointer (&task));
+}
+
+/**
+ * eknc_engine_query_finish:
+ * @self: the engine
+ * @result: the #GAsyncResult that was provided to the callback.
+ * @error: #GError for error reporting.
+ *
+ * Finishes a call to eknc_engine_query. Returns a #EkncQueryResults object.
+ * Throws an error if one occurred.
+ *
+ * Returns: (transfer full): the results object
+ */
+EkncQueryResults *
+eknc_engine_query_finish (EkncEngine *self,
+                          GAsyncResult *result,
+                          GError **error)
+{
+  g_return_val_if_fail (EKNC_IS_ENGINE (self), NULL);
+  g_return_val_if_fail (G_IS_ASYNC_RESULT (result), NULL);
+
+  return g_task_propagate_pointer (G_TASK (result), error);
+}
+
+/**
+ * eknc_engine_get_domain:
+ * @self: the engine
+ * @error: #GError for error reporting.
+ *
+ * Get a #EkncDomain object for the default-app-id. Requires the default-app-id
+ * property to be set.
+ *
+ * Returns the domain if it has already been created, and synchronously creates
+ * one if none exists. Returns an error if one occurred while initializing the
+ * domain.
+ *
+ * Returns: (transfer none): the default domain
+ */
+EkncDomain *
+eknc_engine_get_domain (EkncEngine *self,
+                        GError **error)
+{
+  g_return_val_if_fail (EKNC_IS_ENGINE (self), NULL);
+
+  return eknc_engine_get_domain_for_app (self, self->default_app_id, error);
+}
+
+/**
+ * eknc_engine_get_domain_for_app:
+ * @self: the engine
+ * @app_id: the id of the application to load the object from
+ * @error: #GError for error reporting.
+ *
+ * Get a #EkncDomain object for the a given app id.
+ *
+ * Returns the domain if it has already been created, and synchronously creates
+ * one if none exists. Returns an error if one occurred while initializing the
+ * domain.
+ *
+ * Returns: (transfer none): the domain
```

Comment:
OK, I see now. Maybe a short comment such as `/* takes ownership of domain */` on the `g_hash_table_insert()` call then